constellations/gemini-admin/src/classes/class.js

```diff
@@ -0,0 +1,168 @@
+"use strict";
+require("./class.styl");
+import React from "react";
+import API from "../api.js";
+import { Link } from "react-router-dom";
+import { keys, size } from "lodash";
+import { ClassRow } from "./classRow.js";
+
+export class ClassPage extends React.Component {
+    state = {
+        classes: [],
+        selectedIds: {},
+        unpublished: 0,
+    };
+
+    componentDidMount() {
+        this.fetchData();
+    }
+
+    fetchData() {
+        API.get("api/classes/all").then((res) => {
+            const classes = res.data;
+            const unpublished = classes.filter((c) => !c.publishedAt).length;
+            this.setState({
+                classes: classes,
+                selectedIds: {},
+                unpublished: unpublished,
+            });
+        });
+    }
+
+    onSelectRow = (classId, selected) => {
+        // Switches the checkbox state
+        if (selected) {
+            delete this.state.selectedIds[classId];
+            this.setState({
+                selectedIds: this.state.selectedIds,
+            });
+        } else {
+            this.state.selectedIds[classId] = true;
+            this.setState({
+                selectedIds: this.state.selectedIds,
+            });
+        }
+    };
+
+    onClickSelectAll = () => {
+        // If everything is selected, the Select All button deselects everything
+        if (size(this.state.selectedIds) == this.state.unpublished) {
+            this.state.classes.forEach((c) => {
+                this.onSelectRow(c.classId, true);
+            });
+        } else {
+            this.state.classes.forEach((c) => {
+                if (!c.publishedAt) {
+                    this.onSelectRow(c.classId, false);
+                }
+            });
+        }
+    };
+
+    onClickPublish = () => {
+        const publishList = keys(this.state.selectedIds);
+        console.log("... Publishing Classes ...");
+
+        API.post("api/classes/publish", publishList)
+            .then(() => {
+                console.log("Successfully published classes!");
+                this.fetchData();
+            })
+            .catch((err) => {
+                console.log("Publish failed: " + err);
```

Comment:
For just the failed case, can you do a `window.alert(...)` and display this publish failed message?